drivers/flash/flash_stm32h7x.c

```diff
@@ -102,6 +104,60 @@ static __unused int write_optb(const struct device *dev, uint32_t mask,
 	return 0;
 }
 
+static __unused int write_wp_reg(const struct device *dev, uint32_t mask, uint32_t value,
+				 uint32_t bank)
+{
+	FLASH_TypeDef *regs = FLASH_STM32_REGS(dev);
+	int rc;
+#ifdef CONFIG_SOC_SERIES_STM32H7RSX
+	uintptr_t cur =
+		(uintptr_t)regs + offsetof(FLASH_TypeDef, WRPSR);
+	uintptr_t prg =
+		(uintptr_t)regs + offsetof(FLASH_TypeDef, WRPSRP);
+#else
+	uintptr_t cur =
+		(uintptr_t)regs + offsetof(FLASH_TypeDef, WPSN_CUR1);
+	uintptr_t prg =
+		(uintptr_t)regs + offsetof(FLASH_TypeDef, WPSN_PRG1);
+#endif
+
+#ifdef DUAL_BANK
+	if (bank == 1) {
+		cur = (uint32_t)regs + offsetof(FLASH_TypeDef, WPSN_CUR2);
+		prg = (uint32_t)regs + offsetof(FLASH_TypeDef, WPSN_PRG2);
+	}
+#endif /* DUAL_BANK */
+
+	if (bank >= NUMBER_OF_BANKS) {
+		return -EINVAL;
+	}
+
+	/* Check if already done. */
+	if ((sys_read32(cur) & mask) == value) {
+		return 0;
+	}
+
+	rc = flash_stm32_wait_flash_idle(dev);
+	if (rc < 0) {
+		LOG_ERR("Err flash no idle");
+		return rc;
+	}
+
+	sys_write32((sys_read32(cur) & ~mask) | value, prg);
+	regs->OPTCR |= FLASH_OPTCR_OPTSTART;
+
+	/* Make sure previous write is completed. */
+	barrier_dsync_fence_full();
+
+	rc = flash_stm32_wait_flash_idle(dev);
```

Comment:
From RM0433 Rev 8 p. 175
```
To modify user option bytes, follow the sequence below:

1. Unlock FLASH_OPTCR register as described in Section 4.5.1: FLASH configuration
protection, unless the register is already unlocked.
2. Write the desired new option byte values in the corresponding option registers
(FLASH_XXX_PRG1/2).
3. Set the option byte start change OPTSTART bit to 1 in the FLASH_OPTCR register.
4. Wait until OPT_BUSY bit is cleared.
```

I think we are not waiting for `OPT_BUSY` bit to clear

Reply:
You are so right. I sent #84430 that also fixes that problem for RDP and this PR can reuse added functions. So I'll wait till #84430 is merged

Reply:
Done.